### PR TITLE
Cep inv

### PR DESCRIPTION
**Pull request by kl0u**
R @aljoscha

### CODE REVIEW

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 1]
```diff
@@ -437,72 +425,63 @@ public void add(final ValueTimeWrapper<V> valueTime, final SharedBufferEntry<K,
...
+		private void prune(final long pruningTimestamp, final List<SharedBufferEntry<K, V>> prunedEntries) {
+			Iterator<Map.Entry<ValueTimeWrapper<V>, SharedBufferEntry<K, V>>> it = entries.entrySet().iterator();
+			while (it.hasNext()) {
+				SharedBufferEntry<K, V> entry = it.next().getValue();
 				if (entry.getValueTime().getTimestamp() <= pruningTimestamp) {
 					prunedEntries.add(entry);
-					iterator.remove();
-				} else {
-					continuePruning = false;
+					it.remove();
```
Comment (aljoscha) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: was the early cancel not working before? Or: why was `continuePruning` removed.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 2]
```diff
@@ -519,8 +498,9 @@ public int hashCode() {
 		private final ValueTimeWrapper<V> valueTime;
 		private final Set<SharedBufferEdge<K, V>> edges;
 		private final SharedBufferPage<K, V> page;
+
 		private int referenceCounter;
-		private transient int entryId;
+		private int entryId;
```
Comment (aljoscha) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: Why is this not `transient` anymore?

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 3]
```diff
@@ -578,14 +555,8 @@ private void removeEdges(final List<SharedBufferEntry<K, V>> prunedEntries) {
...
-		public boolean remove() {
-			if (page != null) {
-				page.remove(valueTime);
-
-				return true;
-			} else {
-				return false;
-			}
+		public void remove() {
+			page.remove(valueTime);
```
Comment (aljoscha) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: Did we never need the check before?

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 4]
```diff
@@ -609,22 +580,27 @@ public String toString() {
...
-			} else {
-				return false;
-			}
+					Objects.equals(edges, other.edges);
 		}
 
 		@Override
 		public int hashCode() {
-			return Objects.hash(valueTime, getKey(), referenceCounter, edges);
+			int result = 1;
```
Comment (aljoscha) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: Why not use `Objects.hash()` here anymore?

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 5]
```diff
@@ -542,15 +542,12 @@ public void addEdge(SharedBufferEdge<K, V> edge) {
...
-				for (SharedBufferEntry<K, V> prunedEntry : prunedEntries) {
-					if (prunedEntry == edge.getTarget()) {
-						itor.remove();
-						break;
-					}
+		private void removeEdges(final Set<SharedBufferEntry<K, V>> prunedEntries) {
+			Iterator<SharedBufferEdge<K, V>> it = edges.iterator();
+			while (it.hasNext()) {
+				SharedBufferEdge<K, V> edge = it.next();
+				if (prunedEntries.contains(edge.getTarget())) {
```
Comment (aljoscha) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: Very subtle ... 😅 The `equals()` is used internally in `Set.contains()`.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java (hunk 2 above)
Comment (kl0u) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: we were not serializing it anyway and the class is not `Serializable`.

**Pull request closed without merging**